text/0032-upgrade-bootstrapper.md

```diff
@@ -0,0 +1,152 @@
+
+The `upgrade-bootstrapper` will be a relatively simple Rust binary which will run the `casper-node` binary as a child
+process.  It will take a single optional argument: the path to the `config.toml` for the `casper-node` binary to be run.
+
+Under normal circumstances, the following steps will be performed:
+
+1. the `upgrade-bootstrapper` is started
+2. if the path to the node's `config.toml` was provided via an arg, the `upgrade-bootstrapper` writes the path value to
+   a hard-coded, known location (if writing fails, it panics).  Otherwise it reads the path value from the known
+   location.  If the path cannot be read from the known location, it panics
+3. the `upgrade-bootstrapper` spawns a child process, running the `casper-node` binary in validator mode, passing it the
+   path to its `config.toml` as an arg
+4. while they are running, before the network reaches the upgrade point, the next version of the `casper-node` gets
+   installed, but not run.  The binary is always named `casper-node-next` and its config named `config-next.toml`.  Each
+   gets installed to the same location as the current equivalent files.  The location of the `config.toml` can be
+   retrieved from the known location used by the `upgrade-bootstrapper` at step 2.  An upgrade file for chainspec
+   modifications should also be installed and its path included in the new `config-next.toml`
+5. when the current upgrade activation point is reached and the node has finished all its work in that era, it exits
+   normally with an exit code 0
+6. the `upgrade-bootstrapper` renames
+     * `casper-node-next` to `casper-node` (replaces the old binary)
+     * `config.toml` to `config-old.toml`
+     * `config-next.toml` to `config.toml`
+7. the `upgrade-bootstrapper` runs the new `casper-node` in migrate-data mode (see below for further details), passing
+   it the path to `config.toml` and `config-old.toml` as args.  The node performs any migration of stored data and
+   config options required, then exits normally with an exit code 0
+8. repeat from step 3
+
+Should the child process stop with a non-zero exit code at any stage, the `upgrade-bootstrapper` should itself panic.
+It is up to the user to restart the `upgrade-bootstrapper` manually.
+
+It is assumed that the `casper-node` binary will live at a known location and that installing an upgraded binary will
+place the binary and its config file in a known location.  If we wish to be more flexible about this (particularly the
+location of the `casper-node` binary) we can handle these locations in a similar way to the location of the
+`casper-node`'s `config.toml`.
+
+
+## Reference-level explanation
+
+[reference-level-explanation]: #reference-level-explanation
+
+### Changes to existing `casper-node` (running in validator mode)
+
+We require the following changes to `casper-node`:
+
+* running in validator mode, when we reach the upgrade activation point, exit gracefully
```

Comment:
How does the _already-running_ node knows when it's time for an upgrade? Does it need to be restarted after node operator installs/updates the `upgrade-bootstraper`? Or does it monitor changes to some config file?

Reply:
Andreas had [the same question](https://github.com/CasperLabs/ceps/pull/32#discussion_r555655150) :)

Reply:
So, the process will look as follows:
0. New upgrade activation point defined at Era=10,
1. Active validator finished Era=9, recognizes it's time to upgrade. Exists gracefully. 
2. `upgrade-bootstrapper` recognizes that `casper-node` exit so it will check if there's an upgrade scheduled. To do that, `casper-node` process has to signal this to the `upgrade-bootstraper` somehow – since `upgrade-bootstrapper` doesn't know what era is currently being active.
3. `upgrade-bootstraper` migrates the data. This can take a lot of time.
4. After migration is finished, `upgrade-bootstrapper` will perform the "replacement" of `casper-node` binaries.
5. `upgrade-bootstrapper` starts the new, upgraded binary.
6. `casper-node` starts a new era. Possibly a few minutes after the era started since the migration process takes time.


Does this sound correct?

Reply:
Pretty much.  Minor details are:
* At step 2, the fact that `casper-node` exited with exit code 0 indicates an upgrade is due.  The node has no reason to exit otherwise, and a panic wouldn't return an exit code of 0.
* The migration is performed by the new binary rather than the old, so your steps 3 and 4 should swap places.